core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java

```diff
@@ -677,6 +684,74 @@ Expression translateCast(
     return convert;
   }
 
+  private @Nullable Expression translateCastToDecimal(RelDataType sourceType, Expression operand,
+      RelDataType targetType) {
+    Expression convert = null;
+    switch (sourceType.getSqlTypeName()) {
+    case TINYINT:
+    case SMALLINT:
+    case INTEGER:
+    case BIGINT:
+    case REAL:
+    case FLOAT:
+    case DOUBLE:
+      Expression condition = operand;
+      if(operand instanceof MethodCallExpression) {
+        condition = ((MethodCallExpression) operand).targetExpression;
+      }
+      operand = Expressions.call(BuiltInMethod.STRING_VALUEOF.method, operand);
+      convert = translateCharToDecimal(operand, targetType, condition);
+      break;
+    case CHAR:
+    case VARCHAR:
+      convert = translateCharToDecimal(operand, targetType, operand);
+      break;
+    case DECIMAL:
+      Expression sourceIntDigits =
+          Expressions.constant(sourceType.getPrecision() - sourceType.getScale());
+      Expression targetIntDigits =
+          Expressions.constant(targetType.getPrecision() - targetType.getScale());
+      ConditionalStatement conditionalStatement =
+          Expressions.ifThen(
+              Expressions.andAlso(checkNotNull(operand),
+                  Expressions.greaterThan(sourceIntDigits, targetIntDigits)),
+              Expressions.throw_(
+                  Expressions.new_(RuntimeException.class,
+                      Expressions.constant("numeric field overflow"))));
+      list.add(conditionalStatement);
+      convert = Expressions.call(operand, BuiltInMethod.BIG_DECIMAL_SET_SCALE.method,
+          Expressions.constant(targetType.getScale()), Expressions.constant(RoundingMode.HALF_UP));
+      break;
+    default:
+      break;
+    }
+    return convert;
+  }
+
+  private @NotNull Expression translateCharToDecimal(Expression operand, RelDataType targetType,
```

Comment:
We assume notnull by default, so there's no need for explicit notnull

Reply:
Ok. Thanks you. When the PR is ready I will repair this. 
Actually I need more advire about this PR.
Because When I try to make the cast operator to return the right result(I can confirm the test added in misc.iq have right behavior), then many unit tests throw exception or result have a different scale. So I want to know  how to handle this?

Reply:
> 

guess I don't have a direct reply, but usually I will create another fork of the repo. and compare the behavior simultaneously with the patch and without the patch and adjust correpsondingly.